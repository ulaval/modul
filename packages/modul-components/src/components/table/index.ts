export { MColumnSortDirection, MColumnTable, MColumnTextAlign, MTableSkin } from './table';

