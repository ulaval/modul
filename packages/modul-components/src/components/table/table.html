<table class="m-table"
       :class="['m--is-skin-' + skin, {'m--is-loading': loading, 'm--is-row-highlighted-on-hover': rowHighlightedOnHover}]"
       cellspacing="0">
    <slot name="header">
        <thead>
            <tr>
                <th v-for="(column, index) in columnsInternal"
                    :style="columnWidth(column)"
                    :class="[{ 'm--is-sortable': column.sortable,
                               'm--is-sorted': isColumnSorted(column),
                               'm--is-text-align-center': isColumnTextAlignCenter(column),
                               'm--is-text-align-left': isColumnTextAlignLeft(column),
                               'm--is-text-align-right': isColumnTextAlignRight(column),
                               'm--is-initial-sort': column.isInitialSort }, column.class]"
                    :key="column.id"
                    @click="sort(column)"
                    scope="col">
                    <slot :name="'header.' + column.dataProp"
                          :column="column">
                        <div v-if="column.sortable"
                             class="m-table__header-wrap">
                            <span class="m-table__sortable-icon-wrap">
                                <span class="m-table__header-name"
                                      v-html="column.title"></span>
                                <m-icon-button v-if="column.sortable"
                                               class="m-table__sortable-icon"
                                               :class="[getColumnSortDirectionClass(column)]"
                                               :disabled="loading"
                                               :icon-name="getColumnSortIcon(column)"
                                               skin="bold"
                                               button-size="14px"
                                               icon-size="12px"
                                               :title="i18nSort"
                                               @click.stop="sort(column)"></m-icon-button>
                            </span>

                        </div>
                        <template v-else>
                            <span class="m-table__header-name"
                                  v-html="column.title"></span>
                        </template>
                    </slot>

                </th>

            </tr>
        </thead>
    </slot>
    <slot name="body">
        <tbody>
            <tr class="m-table__placeholder"
                v-if="isEmpty || loading">
                <div class="m-table__placeholder-container"
                     :style="{position: placeholderPositionType, width: widthPlaceholder}">
                    <m-progress v-if="loading"
                                :indeterminate="true"
                                :border-radius="false"></m-progress>
                    <div class="m-table__placeholder-message">
                        <slot name="empty"
                              v-if="isEmpty">
                            <p>{{ i18nEmptyTable }}</p>
                        </slot>
                        <slot name="loading"
                              v-if="loading">
                            <p>{{ i18nLoading }}</p>
                            <p class="m-table__placeholder-precision">{{ i18nPleaseWait }}</p>
                        </slot>
                    </div>
                </div>
            </tr>
            <tr v-else
                v-for="(row, index) in rows"
                :key="index">
                <slot name="row"
                      :data="row">
                    <td v-for="column in columnsInternal"
                        :class="[{ 'm--is-sortable': column.sortable,
                                  'm--is-text-align-center': isColumnTextAlignCenter(column),
                                  'm--is-text-align-left': isColumnTextAlignLeft(column),
                                  'm--is-text-align-right': isColumnTextAlignRight(column) }, column.class]"
                        :key="'body.' + column.id">
                        <slot :name="'body.' + column.dataProp"
                              :data="row">{{ row[column.dataProp] }}</slot>
                    </td>
                </slot>
            </tr>
        </tbody>
    </slot>

    <tfoot v-if="$slots.footer">
        <tr>
            <slot name="footer"></slot>
        </tr>
    </tfoot>
</table>
