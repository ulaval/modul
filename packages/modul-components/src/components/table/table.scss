@import 'commons';

$m-sortable-icon-size: 14px;

@mixin m-table-sortable-icon-select() {
    opacity: 1;

    &.m--is-sort-desc {
        transform: rotate(180deg);
    }
}

.m-table {
    position: relative;
    border-collapse: collapse;

    &.m--is-loading {
        thead {
            th {
                &.m--is-sortable {
                    .m-table__header-name {
                        cursor: default;
                    }
                }
            }
        }
    }

    &.m--is-row-highlighted-on-hover {
        tbody {
            tr {
                transition: background $m-transition-duration ease;
            }

            tr:hover:not(.m-table__message):not(.m-table__loading) {
                background-color: $m-color--interactive-lightest;
            }
        }
    }

    thead {
        text-align: left;

        tr {
            background-color: rgba($m-color--grey-light, 0.3);
        }

        th {
            color: $m-color--grey;
            font-weight: $m-font-weight--semi-bold;

            &.m--is-sortable {
                cursor: pointer;

                &:hover {
                    color: $m-color--grey-darker;

                    .m-table__sortable-icon {
                        color: $m-color--grey;
                        opacity: 1;
                    }
                }

                &.m--is-sorted {
                    color: $m-color--grey-darker;

                    .m-table__sortable-icon {
                        color: $m-color--grey-darker;

                        @include m-table-sortable-icon-select();
                    }
                }

                &.m--is-initial-sort {
                    .m-table__sortable-icon {
                        transition: transform 0s, opacity $m-transition-duration ease;
                    }
                }
            }

            &.m--is-text-align-center {
                text-align: center;

                .m-table__header-wrap {
                    justify-content: center;
                }
            }

            &.m--is-text-align-left {
                text-align: left;

                .m-table__header-wrap {
                    justify-content: flex-start;
                }
            }

            &.m--is-text-align-right {
                text-align: right;

                .m-table__header-wrap {
                    justify-content: flex-end;
                }
            }
        }
    }

    tbody {
        color: $m-color--grey-darker;
    }

    th,
    td {
        box-sizing: content-box;
        padding-top: $m-spacing;
        padding-bottom: $m-spacing;

        &:first-child {
            padding-left: $m-spacing;
        }

        &:last-child {
            padding-right: $m-spacing;
        }

        &:not(:first-child) {
            padding-left: $m-spacing--s;
        }

        &:not(:last-child) {
            padding-right: $m-spacing--s;
        }

        &.m--is-text-align-center {
            text-align: center;

            > * {
                text-align: center;
            }
        }

        &.m--is-text-align-left {
            text-align: left;

            > * {
                text-align: left;
            }
        }

        &.m--is-text-align-right {
            text-align: right;

            > * {
                text-align: right;
            }
        }
    }

    th {
        &:first-child {
            &.m--is-sortable {
                &.m--is-text-align-center {
                    padding-left: calc(#{$m-sortable-icon-size} + #{$m-spacing--xs} + #{$m-spacing});
                }
            }
        }

        &:last-child {
            &.m--is-sortable {
                padding-right: calc(#{$m-sortable-icon-size} + #{$m-spacing--xs} + #{$m-spacing});
            }
        }

        &:not(:first-child) {
            &.m--is-sortable {
                &.m--is-text-align-center {
                    padding-left: calc(#{$m-sortable-icon-size} + #{$m-spacing--xs} + #{$m-spacing--s});
                }
            }
        }

        &:not(:last-child) {
            &.m--is-sortable {
                padding-right: calc(#{$m-sortable-icon-size} + #{$m-spacing--xs} + #{$m-spacing--s});
            }
        }
    }

    td {
        &:last-child {
            &.m--is-sortable.m--is-text-align-right {
                padding-right: calc(#{$m-sortable-icon-size} + #{$m-spacing--xs} + #{$m-spacing});
            }
        }

        &:not(:last-child) {
            &.m--is-sortable.m--is-text-align-right {
                padding-right: calc(#{$m-sortable-icon-size} + #{$m-spacing--xs} + #{$m-spacing--s});
            }
        }
    }

    &__header-wrap {
        display: flex;
        align-items: center;
    }

    &__sortable-icon-wrap {
        position: relative;
        display: flex;
        align-items: center;
    }

    &__header-name {
        position: relative;
        z-index: 1;
    }

    .m-table__sortable-icon {
        transition: transform $m-transition-duration ease, opacity $m-transition-duration ease;
        position: absolute;
        right: calc((#{$m-sortable-icon-size} + #{$m-spacing--xs}) * -1);
        z-index: 0;
        opacity: 0;

        &:focus {
            color: $m-color--text;
            opacity: 1;
        }
    }

    &__loading.m-table__loading > td {
        padding: 0;
        height: auto;
    }

    &__placeholder,
    &__placeholder-container {
        height: 260px;
    }

    &__placeholder-container {
        position: sticky;
        right: 0;
        left: 0;
        display: flex;
        flex-direction: column;
        background-color: $m-color--grey-lightest;
    }

    &__placeholder-message {
        display: flex;
        flex-direction: column;
        justify-content: center;
        align-items: center;
        height: 100%;
        min-width: 200px;
        padding: $m-spacing;
        color: $m-color--grey;
        font-weight: $m-font-weight--semi-bold;
        font-size: $m-font-size--l;
        text-align: center;

        > :first-child {
            margin-top: 0;
        }
    }

    &__placeholder-precision {
        margin: $m-spacing--xs 0 0 0;
        font-size: $m-font-size;
        font-weight: $m-font-weight--light;
    }
}

.m-table.m--is-skin-regular {
    tbody {
        tr {
            &:nth-child(even) {
                background-color: $m-color--grey-lightest;
            }
        }
    }
}

.m-table.m--is-skin-simple {
    tbody {
        tr:not(:first-child) {
            border-top: 1px solid $m-color--grey-lighter;
        }
    }
}
