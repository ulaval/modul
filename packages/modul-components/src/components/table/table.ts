import { PluginObject } from 'vue';
import Component from 'vue-class-component';
import { Emit, Prop } from 'vue-property-decorator';
import { Enums } from '../../utils/enums/enums';
import { ModulVue } from '../../utils/vue/vue';
import { TABLE_NAME } from '../component-names';
import ProgressPlugin from '../progress/progress';
import WithRender from './table.html?style=./table.scss';


export enum MTableSkin {
    Regular = 'regular',
    Simple = 'simple'
}

export enum MColumnSortDirection {
    None = 0,
    Asc = 1,
    Dsc = -1
}

export enum MColumnTextAlign {
    Center = 'center',
    Left = 'left',
    Right = 'right'
}

export interface MColumnTable {
    id: string;
    title: string;
    dataProp: string;
    width?: string;
    sortable?: boolean;
    enableUnsort?: boolean;
    textAlign?: MColumnTextAlign;
    class?: string;
    sortDirection?: MColumnSortDirection;
    defaultSortDirection?: MColumnSortDirection;
}

interface MColumnTableInternal extends MColumnTable {
    isInitialSort?: boolean;
}

@WithRender
@Component
export class MTable extends ModulVue {

    @Prop({
        default: MTableSkin.Regular,
        validator: value => Enums.toValueArray(MTableSkin).includes(value)
    })
    public skin: MTableSkin;

    @Prop({ default: () => [] })
    public columns: MColumnTable[];

    @Prop({ default: () => [] })
    public rows: any[];

    @Prop({ default: false })
    public loading: boolean;

    @Prop({ default: true })
    public rowHighlightedOnHover: boolean;

    public i18nEmptyTable: string = this.$i18n.translate('m-table:no-data');
    public i18nLoading: string = this.$i18n.translate('m-table:loading');
    public i18nPleaseWait: string = this.$i18n.translate('m-table:please-wait');
    public i18nSort: string = this.$i18n.translate('m-table:sort');

    @Emit('add')
    private onAdd(): void {
    }

    @Emit('sort-applied')
    private emitSortApplied(columnTable: MColumnTable): void { }

    public get isEmpty(): boolean {
        return this.rows.length === 0 && !this.loading;
    }

    public sort(columnTable: MColumnTableInternal): void {
        if (this.loading || !columnTable.sortable) {
            return;
        }

        if (typeof columnTable.sortDirection === 'undefined') {
            columnTable.sortDirection = MColumnSortDirection.None;
        }

        this.columnsInternal.forEach(c => {
            if (c !== columnTable) {
                c.sortDirection = MColumnSortDirection.None;
            }
        });

        switch (columnTable.sortDirection) {
            case MColumnSortDirection.None:
                columnTable.sortDirection = columnTable.defaultSortDirection ? columnTable.defaultSortDirection : MColumnSortDirection.Asc;
                columnTable.isInitialSort = true;
                break;
            case MColumnSortDirection.Asc:
                if (columnTable.enableUnsort) {
                    columnTable.sortDirection = columnTable.defaultSortDirection === MColumnSortDirection.Dsc ? MColumnSortDirection.None : MColumnSortDirection.Dsc;
                } else {
                    columnTable.sortDirection = MColumnSortDirection.Dsc;
                }
                columnTable.isInitialSort = false;
                break;
            case MColumnSortDirection.Dsc:
                if (columnTable.enableUnsort) {
                    columnTable.sortDirection = (columnTable.defaultSortDirection === MColumnSortDirection.Asc || columnTable.defaultSortDirection === undefined) ? MColumnSortDirection.None : MColumnSortDirection.Asc;
                } else {
                    columnTable.sortDirection = MColumnSortDirection.Asc;
                }
                columnTable.isInitialSort = false;
                break;
        }

        this.emitSortApplied(columnTable);
    }

    public isColumnSorted(columnTable: MColumnTable): boolean {
        return columnTable.sortDirection === MColumnSortDirection.Asc || columnTable.sortDirection === MColumnSortDirection.Dsc;
    }

    public isColumnTextAlignCenter(columnTable: MColumnTable): boolean {
        return columnTable.textAlign === MColumnTextAlign.Center;
    }

    public isColumnTextAlignLeft(columnTable: MColumnTable): boolean {
        return columnTable.textAlign === MColumnTextAlign.Left;
    }

    public isColumnTextAlignRight(columnTable: MColumnTable): boolean {
        return columnTable.textAlign === MColumnTextAlign.Right;
    }

    public getColumnSortDirectionClass(columnTable: MColumnTable): string | undefined {
        switch (columnTable.sortDirection) {
            case MColumnSortDirection.Asc:
                return 'm--is-sort-asc';
            case MColumnSortDirection.Dsc:
                return 'm--is-sort-desc';
            default:
                if (columnTable.defaultSortDirection) {
                    return columnTable.defaultSortDirection === MColumnSortDirection.Asc ? 'm--is-sort-asc' : 'm--is-sort-desc';
                } else {
                    return undefined;
                }
        }
    }

    public getColumnSortIcon(columnTable: MColumnTable): string {
        if (columnTable.sortDirection) {
            return 'm-svg__arrow-thin--up'; // CSS rotate transform animation takes care of the arrow position
        } else {
            return columnTable.defaultSortDirection === MColumnSortDirection.Dsc ? 'm-svg__arrow-thin--down' : 'm-svg__arrow-thin--up';
        }
    }

    public columnWidth(col: MColumnTable): { width: string } | '' {
        return col.width ? { width: col.width } : '';
    }

    get columnsInternal(): MColumnTableInternal[] {
        return this.columns.map((c: MColumnTable) => ({ ...c }));
    }
}

const TablePlugin: PluginObject<any> = {
    install(v, options): void {
        v.use(ProgressPlugin);
        v.component(TABLE_NAME, MTable);
    }
};

export default TablePlugin;
