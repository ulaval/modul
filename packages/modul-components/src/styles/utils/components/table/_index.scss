@import 'table.mixin';

@include mu-table();
