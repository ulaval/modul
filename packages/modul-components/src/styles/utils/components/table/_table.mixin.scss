@mixin mu-table {
    .mu-table {
        @include m-table();
    }
}
