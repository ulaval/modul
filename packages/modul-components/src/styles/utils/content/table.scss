.mu-table {
    @include m-table();
}
