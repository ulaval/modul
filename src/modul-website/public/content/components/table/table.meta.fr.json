{
    "name": "Tableau",
    "url": "tableau",
    "category": "tables",
    "type": "components",
    "beta": false,
    "visible": true,
    "preview": "components/table/table.preview.fr.md",
    "components": [
        "m-table",
        "m-table-header"
    ]
}
