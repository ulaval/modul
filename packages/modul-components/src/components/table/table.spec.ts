import { shallowMount, Wrapper } from '@vue/test-utils';
import Vue from 'vue';
import { MColumnSortDirection, MColumnTable, MTable } from './table';

let slots: any = {};
const SLOT_EMPTY: string = '<p>EMPTY</p>';
const SLOT_HEADER: string = '<thead><tr><th>SLOT THEAD</th></tr></thead>';
const SLOT_TH: string = 'SLOT TH';
const SLOT_BODY: string = '<tbody><tr><td>SLOT BODY</td></tr></tbody>';
const SLOT_TD: string = 'SLOT TD';
const SLOT_FOOTER: string = '<td>SLOT FOOTER</td>';

let rows: any[] = [];

const columns: MColumnTable[] = [
    { id: 'a', title: 'A', dataProp: 'a', width: '10%', sortable: true, sortDirection: MColumnSortDirection.None },
    { id: 'b', title: 'B', dataProp: 'b', sortDirection: MColumnSortDirection.None }
];

const EMPTY_DATA: any[] = [];
const DATA: any[] = [
    { id: '1', a: 'a1', b: 'b1' },
    { id: '2', a: 'a2', b: 'b2' }
];

let wrapper: Wrapper<MTable>;

const initializeShallowWrapper: any = () => {
    wrapper = shallowMount(MTable, {
        propsData: {
            rows,
            columns
        },
        slots
    });
};

describe(`MTable`, () => {

    describe(`Given an empty table`, () => {

        beforeEach(() => {
            rows = EMPTY_DATA;
        });

        it(`Then should be empty`, () => {
            initializeShallowWrapper();

            expect(wrapper.vm.isEmpty).toBeTruthy();
        });

        describe(`When loading`, () => {
            beforeEach(() => {
                initializeShallowWrapper();
                wrapper.setProps({ loading: true });
            });

            it(`Then should not be empty`, () => {
                expect(wrapper.vm.isEmpty).toBeFalsy();
            });


        });



    });

    describe(`Given a table full of data`, () => {

        beforeEach(() => {
            rows = DATA;
        });

        it(`Then should not be empty`, () => {
            initializeShallowWrapper();

            expect(wrapper.vm.isEmpty).toBeFalsy();
        });

        describe(`When we don't use slots`, () => {

            describe(`When a column is sortable`, () => {
                it(`Should show a button with the label on it`, async () => {
                    const icon: Wrapper<Vue> = wrapper.find('.m-table__sortable-icon');

                    expect(icon.exists()).toBeTruthy();
                });

                describe(`When the arrow is clicked`, () => {
                    it(`Should emit a action event`, () => {
                        wrapper.find('.m-table__sortable-icon').trigger('click');

                        expect(wrapper.emitted('sort-applied')).toBeTruthy();
                    });
                });
            });
        });


    });

    describe(`Given a column`, () => {

        describe(`When the width is specified`, () => {
            it(`Then should be the right width`, () => {
                initializeShallowWrapper();

                expect(wrapper.vm.columnWidth(columns[0])).toEqual({ width: '10%' });
            });
        });

        describe(`When the width is undefined`, () => {
            it(`Then should be the right width`, () => {
                initializeShallowWrapper();

                expect(wrapper.vm.columnWidth(columns[1])).toEqual('');
            });
        });

    });

});
