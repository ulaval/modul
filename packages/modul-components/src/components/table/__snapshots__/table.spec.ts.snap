// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MTable Given a table full of data When using a footer Then should render correctly 1`] = `
<table cellspacing="0" class="m-table m--is-skin-regular m--is-row-highlighted-on-hover">
  <thead>
    <tr>
      <th scope="col" class="m--is-sortable" style="width:10%;">
        <div class="m-table__header-wrap">
          <div class="m-table__sortable-icon-wrap">
            <div class="m-table__header-name">A</div>
            <m-icon-button icon-name="m-svg__arrow-thin--up" skin="bold" button-size="14px" icon-size="12px" title="m-table:sort" class="m-table__sortable-icon"></m-icon-button>
          </div>
        </div>
      </th>
      <th scope="col">
        <div class="m-table__header-name">B</div>
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td class="m--is-sortable">a1</td>
      <td>b1</td>
    </tr>
    <tr>
      <td class="m--is-sortable">a2</td>
      <td>b2</td>
    </tr>
  </tbody>
  <tfoot>
    <tr>
      <td>SLOT FOOTER</td>
    </tr>
  </tfoot>
</table>
`;

exports[`MTable Given a table full of data When using slot to replace a td Then should render correctly 1`] = `
<table cellspacing="0" class="m-table m--is-skin-regular m--is-row-highlighted-on-hover">
  <thead>
    <tr>
      <th scope="col" class="m--is-sortable" style="width:10%;">
        <div class="m-table__header-wrap">
          <div class="m-table__sortable-icon-wrap">
            <div class="m-table__header-name">A</div>
            <m-icon-button icon-name="m-svg__arrow-thin--up" skin="bold" button-size="14px" icon-size="12px" title="m-table:sort" class="m-table__sortable-icon"></m-icon-button>
          </div>
        </div>
      </th>
      <th scope="col">
        <div class="m-table__header-name">B</div>
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td class="m--is-sortable">SLOT TD</td>
      <td>b1</td>
    </tr>
    <tr>
      <td class="m--is-sortable">SLOT TD</td>
      <td>b2</td>
    </tr>
  </tbody>
</table>
`;

exports[`MTable Given a table full of data When using slot to replace a th Then should render correctly 1`] = `
<table cellspacing="0" class="m-table m--is-skin-regular m--is-row-highlighted-on-hover">
  <thead>
    <tr>
      <th scope="col" class="m--is-sortable" style="width:10%;">
        <div class="m-table__header-wrap">
          <div class="m-table__sortable-icon-wrap">SLOT TH <m-icon-button icon-name="m-svg__arrow-thin--up" skin="bold" button-size="14px" icon-size="12px" title="m-table:sort" class="m-table__sortable-icon"></m-icon-button>
          </div>
        </div>
      </th>
      <th scope="col">
        <div class="m-table__header-name">B</div>
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td class="m--is-sortable">a1</td>
      <td>b1</td>
    </tr>
    <tr>
      <td class="m--is-sortable">a2</td>
      <td>b2</td>
    </tr>
  </tbody>
</table>
`;

exports[`MTable Given a table full of data When using slot to replace the header Then should render correctly 1`] = `
<table cellspacing="0" class="m-table m--is-skin-regular m--is-row-highlighted-on-hover">
  <thead>
    <tr>
      <th>SLOT THEAD</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td class="m--is-sortable">a1</td>
      <td>b1</td>
    </tr>
    <tr>
      <td class="m--is-sortable">a2</td>
      <td>b2</td>
    </tr>
  </tbody>
</table>
`;

exports[`MTable Given a table full of data When using slot to replace the tbody Then should render correctly 1`] = `
<table cellspacing="0" class="m-table m--is-skin-regular m--is-row-highlighted-on-hover">
  <thead>
    <tr>
      <th scope="col" class="m--is-sortable" style="width:10%;">
        <div class="m-table__header-wrap">
          <div class="m-table__sortable-icon-wrap">
            <div class="m-table__header-name">A</div>
            <m-icon-button icon-name="m-svg__arrow-thin--up" skin="bold" button-size="14px" icon-size="12px" title="m-table:sort" class="m-table__sortable-icon"></m-icon-button>
          </div>
        </div>
      </th>
      <th scope="col">
        <div class="m-table__header-name">B</div>
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>SLOT BODY</td>
    </tr>
  </tbody>
</table>
`;

exports[`MTable Given a table full of data When we don't use slots Then should render correctly 1`] = `
<table cellspacing="0" class="m-table m--is-skin-regular m--is-row-highlighted-on-hover">
  <thead>
    <tr>
      <th scope="col" class="m--is-sortable" style="width:10%;">
        <div class="m-table__header-wrap">
          <div class="m-table__sortable-icon-wrap">
            <div class="m-table__header-name">A</div>
            <m-icon-button icon-name="m-svg__arrow-thin--up" skin="bold" button-size="14px" icon-size="12px" title="m-table:sort" class="m-table__sortable-icon"></m-icon-button>
          </div>
        </div>
      </th>
      <th scope="col">
        <div class="m-table__header-name">B</div>
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td class="m--is-sortable">a1</td>
      <td>b1</td>
    </tr>
    <tr>
      <td class="m--is-sortable">a2</td>
      <td>b2</td>
    </tr>
  </tbody>
</table>
`;

exports[`MTable Given an empty table When a custom slot is given Then should render correctly 1`] = `
<table cellspacing="0" class="m-table m--is-skin-regular m--is-row-highlighted-on-hover">
  <thead>
    <tr>
      <th scope="col" class="m--is-sortable" style="width:10%;">
        <div class="m-table__header-wrap">
          <div class="m-table__sortable-icon-wrap">
            <div class="m-table__header-name">A</div>
            <m-icon-button icon-name="m-svg__arrow-thin--up" skin="bold" button-size="14px" icon-size="12px" title="m-table:sort" class="m-table__sortable-icon"></m-icon-button>
          </div>
        </div>
      </th>
      <th scope="col">
        <div class="m-table__header-name">B</div>
      </th>
    </tr>
  </thead>
  <tbody>
    <tr class="m-table__placeholder">
      <td class="m-table__placeholder-container" style="position:absolute;width:100%;">
        <div class="m-table__placeholder-message">
          <p>EMPTY</p>
        </div>
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`MTable Given an empty table When loading Then should render correctly 1`] = `
<table cellspacing="0" class="m-table m--is-skin-regular m--is-loading m--is-row-highlighted-on-hover">
  <thead>
    <tr>
      <th scope="col" class="m--is-sortable" style="width:10%;">
        <div class="m-table__header-wrap">
          <div class="m-table__sortable-icon-wrap">
            <div class="m-table__header-name">A</div>
            <m-icon-button disabled="disabled" icon-name="m-svg__arrow-thin--up" skin="bold" button-size="14px" icon-size="12px" title="m-table:sort" class="m-table__sortable-icon"></m-icon-button>
          </div>
        </div>
      </th>
      <th scope="col">
        <div class="m-table__header-name">B</div>
      </th>
    </tr>
  </thead>
  <tbody>
    <tr class="m-table__placeholder">
      <td class="m-table__placeholder-container" style="position:absolute;width:100%;">
        <m-progress indeterminate="indeterminate"></m-progress>
        <div class="m-table__placeholder-message">
          <p>m-table:loading</p>
          <p class="m-table__placeholder-precision">m-table:please-wait</p>
        </div>
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`MTable Given an empty table When we use the default template Then should render correctly 1`] = `
<table cellspacing="0" class="m-table m--is-skin-regular m--is-row-highlighted-on-hover">
  <thead>
    <tr>
      <th scope="col" class="m--is-sortable" style="width:10%;">
        <div class="m-table__header-wrap">
          <div class="m-table__sortable-icon-wrap">
            <div class="m-table__header-name">A</div>
            <m-icon-button icon-name="m-svg__arrow-thin--up" skin="bold" button-size="14px" icon-size="12px" title="m-table:sort" class="m-table__sortable-icon"></m-icon-button>
          </div>
        </div>
      </th>
      <th scope="col">
        <div class="m-table__header-name">B</div>
      </th>
    </tr>
  </thead>
  <tbody>
    <tr class="m-table__placeholder">
      <td class="m-table__placeholder-container" style="position:absolute;width:100%;">
        <div class="m-table__placeholder-message">
          <p>m-table:no-data</p>
        </div>
      </td>
    </tr>
  </tbody>
</table>
`;
